content/posts/2022-05-13-calculating-type-sets.markdown

```diff
@@ -0,0 +1,498 @@
+
+## Basic interfaces as type sets
+
+Interfaces in Go have always described sets of types. For example, the interface
+
+```go
+type S interface {
+    X()
+    Y()
+    Z()
+}
+```
+
+is “the set of all types which have a method `X()` and a method `Y()` and a method `Z()`”.
+
+We also had a way to express set intersection, using [interface embedding](https://go.dev/ref/spec#Embedded_interfaces):
+
+
+```go
+type S interface { X() }
+type T interface { Y() }
+type U interface {
+    S
+    T
+}
+```
+
+This expresses `T=S∩T` as an interface. We can also view the predicate “has a
+method `X()`” as a boolean variable and think of this as the formula `X∧Y`.
+
+Surprisingly, there is also a simple form of negation. It happens implicitly,
+because a type can not have two methods with the same name, but different
+types. So, implicitly, if a type has a method `X()`, it does *not* have a
+method `X() int`, for example[^5]:
+
+```go
+type X interface { X() }
+type NotX interface{ X() int }
+```
+
+This meant that you could express interfaces which could never be satisfied by
+any type (i.e. which described an empty type set):
+
+```go
+interface{ X; NotX }
+```
+
+[The compiler rejects such interfaces](https://go.dev/play/p/r4kpXNynscX).
+
+So, we pretty much had a language to build boolean formulas, where every
+formula had the form
+
+    X∧Y∧¬Z
+
+Checking if a type implements an interface meant to check if it is in the set
+described by that interfaces. Which meant checking if the corresponding formula
+is satisfied.
+
+And checking if an interface is invalid meant checking if the set it describes
+is empty. Which meant checking if the formula is *satisfiable*. This was easy,
+though. If you look closely, those formulas are in DNF and as we noted above,
+solving SAT on a formula given in DNF is easy.
+
+## Adding unions
+
+Go 1.18 extends the interface syntax. For our purposes, the important addition
+is the `|` operator:
+
+```go
+type S interface{
+    A | B
+}
+```
+
+This represents the set of all types which are in the *union* of the type sets
+`A` and `B` - that is, the set of all types which are in `A` *or* in `B` (or
+both).
+
+This means our language of expressible formulas now also includes a
+`∨`-operator. What's more, the form of our formula is now a *Conjunctive*
+normal form - every line is a term of `∨` and the lines are connected by `∧`:
+
+```go
+type X interface { X() }
+type NotX interface{ X() int }
+type Y interface { Y() }
+type NotY interface{ Y() int }
+type Z interface { Z() }
+type NotZ interface{ Z() int }
+
+// (¬X ∨ Z) ∧ (Y ∨ Z) ∧ (X ∨ ¬Y)
+type S interface {
+    NotX | Z
+    Y | Z
+    X | NotY
+}
+```
+
+There is one detail which becomes important here. There are types which
+*neither* implement `X` *nor* do they implement `NotX` - if they don't have an
+`X` method at all, for example. If we want to really prove that calculating
+these new type sets is NP complete, we need to fix that. We want that a type
```

Comment:
s/want/need/ ?

Also, why exactly _is_ this a problem for determining the type set?

Reply:
I'm not sure I understand the question.

Throughout the article, I kind of talk about "calculating the type set" and "checking if a type set is empty" as sort of the same thing. Is that what you are asking? Because that does probably need clarification (I'm not sure where and how).